File: flutter/test/deep_link_test.dart
import 'package:busha_pay/busha_pay_config.dart';
import 'package:busha_pay/src/deep_link.dart';
import 'package:busha_pay/src/sdk.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_test/flutter_test.dart';

const _config = BushaPayConfig(
  quoteAmount: '10000',
  quoteCurrency: 'NGN',
  targetCurrency: 'NGN',
  sourceCurrency: 'USDT',
);

void main() {
  tearDown(() => debugDefaultTargetPlatformOverride = null);

  group('bushaAppScheme', () {
    test('iOS live → co.busha.apple', () {
      debugDefaultTargetPlatformOverride = TargetPlatform.iOS;
      expect(bushaAppScheme(BushaEnvironment.live), 'co.busha.apple');
    });

    test('iOS sandbox → co.busha.boro.development', () {
      debugDefaultTargetPlatformOverride = TargetPlatform.iOS;
      expect(bushaAppScheme(BushaEnvironment.sandbox), 'co.busha.boro.development');
    });

    test('Android live → co.busha.android', () {
      debugDefaultTargetPlatformOverride = TargetPlatform.android;
      expect(bushaAppScheme(BushaEnvironment.live), 'co.busha.android');
    });

    test('Android sandbox → co.busha.android.development', () {
      debugDefaultTargetPlatformOverride = TargetPlatform.android;
      expect(bushaAppScheme(BushaEnvironment.sandbox), 'co.busha.android.development');
    });

    test('macOS / unsupported → null', () {
      debugDefaultTargetPlatformOverride = TargetPlatform.macOS;
      expect(bushaAppScheme(BushaEnvironment.live), isNull);
    });
  });

  group('buildBushaAppDeepLink', () {
    test('returns null when scheme is null', () {
      final uri = buildBushaAppDeepLink(scheme: null, config: _config, publicKey: 'pk', callbackUrl: 'cb://cb');
      expect(uri, isNull);
    });

    test('builds the Uri host + path', () {
      final uri = buildBushaAppDeepLink(
        scheme: 'co.busha.apple',
        config: _config,
        publicKey: 'pub_test',
        callbackUrl: 'co.example.busha-pay://callback',
      )!;
      expect(uri.scheme, 'co.busha.apple');
      expect(uri.host, 'busha.co');
      expect(uri.path, '/pay');
    });

    test('embeds the required query parameters', () {
      final uri = buildBushaAppDeepLink(
        scheme: 'co.busha.android',
        config: _config,
        publicKey: 'pub_test',
        callbackUrl: 'co.example.busha-pay://callback',
      )!;
      expect(uri.queryParameters['public_key'], 'pub_test');
      expect(uri.queryParameters['quote_amount'], '10000');
      expect(uri.queryParameters['quote_currency'], 'NGN');
      expect(uri.queryParameters['target_currency'], 'NGN');
      expect(uri.queryParameters['callback_url'], 'co.example.busha-pay://callback');
    });

    test('omits reference when not provided', () {
      final uri = buildBushaAppDeepLink(
        scheme: 'co.busha.apple',
        config: _config,
        publicKey: 'pk',
        callbackUrl: 'cb://cb',
      )!;
      expect(uri.queryParameters.containsKey('reference'), isFalse);
    });

    test('includes reference when provided', () {
      const c = BushaPayConfig(
        quoteAmount: '10000',
        quoteCurrency: 'NGN',
        targetCurrency: 'NGN',
        sourceCurrency: 'USDT',
        reference: 'ref-1',
      );
      final uri = buildBushaAppDeepLink(scheme: 'co.busha.apple', config: c, publicKey: 'pk', callbackUrl: 'cb://cb')!;
      expect(uri.queryParameters['reference'], 'ref-1');
    });
  });
}

File: android/pay-android/src/test/kotlin/co/busha/pay/internal/DeepLinkTest.kt
package co.busha.pay.internal

import co.busha.pay.BushaEnvironment
import co.busha.pay.BushaPayCancelledReason
import co.busha.pay.BushaPayConfig
import co.busha.pay.BushaPayResult
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class DeepLinkTest {

    private val config = BushaPayConfig(
        quoteAmount = "10000",
        quoteCurrency = "NGN",
        targetCurrency = "NGN",
        sourceCurrency = "USDT",
    )

    @Test
    fun isWebScheme_acceptsWebSchemesCaseInsensitively() {
        assertTrue(isWebScheme("http"))
        assertTrue(isWebScheme("https"))
        assertTrue(isWebScheme("HTTPS"))
        assertTrue(isWebScheme("about"))
        assertTrue(isWebScheme("data"))
        assertTrue(isWebScheme("blob"))
    }

    @Test
    fun isWebScheme_rejectsNonWebSchemes() {
        assertFalse(isWebScheme("mailto"))
        assertFalse(isWebScheme("co.busha.android"))
        assertFalse(isWebScheme("tel"))
    }

    @Test
    fun bushaAppScheme_mapsEnvironments() {
        assertEquals("co.busha.android", bushaAppScheme(BushaEnvironment.LIVE))
        assertEquals(
            "co.busha.android.development",
            bushaAppScheme(BushaEnvironment.SANDBOX),
        )
    }

    @Test
    fun buildBushaAppDeepLink_buildsHostAndPath() {
        val url = buildBushaAppDeepLink(
            scheme = "co.busha.android",
            config = config,
            publicKey = "pub_x",
            callbackUrl = "co.example.app.busha-pay://callback",
        )
        assertTrue(url.startsWith("co.busha.android://busha.co/pay?"))
    }

    @Test
    fun buildBushaAppDeepLink_embedsRequiredParams() {
        val url = buildBushaAppDeepLink(
            scheme = "co.busha.android",
            config = config,
            publicKey = "pub_test",
            callbackUrl = "co.example.app.busha-pay://callback",
        )
        assertTrue(url.contains("public_key=pub_test"))
        assertTrue(url.contains("quote_amount=10000"))
        assertTrue(url.contains("quote_currency=NGN"))
        assertTrue(url.contains("target_currency=NGN"))
        assertTrue(url.contains("callback_url=co.example.app.busha-pay%3A%2F%2Fcallback"))
    }

    @Test
    fun buildBushaAppDeepLink_omitsReferenceWhenAbsent() {
        val url = buildBushaAppDeepLink("co.busha.android", config, "pk", "cb://cb")
        assertFalse(url.contains("reference="))
    }

    @Test
    fun buildBushaAppDeepLink_includesReferenceWhenSet() {
        val url = buildBushaAppDeepLink(
            "co.busha.android",
            config.copy(reference = "ref-1"),
            "pk",
            "cb://cb",
        )
        assertTrue(url.contains("reference=ref-1"))
    }

    @Test
    fun parseCallback_completedYieldsSuccess() {
        val r = parseCallback(
            "co.x.busha-pay://callback?status=completed&paymentRequestId=PAYR_1",
        )
        assertTrue(r is BushaPayResult.Success)
        assertEquals("PAYR_1", (r as BushaPayResult.Success).paymentId)
    }

    @Test
    fun parseCallback_cancelledYieldsRejectedCancellationWithPaymentId() {
        val r = parseCallback(
            "co.x.busha-pay://callback?status=cancelled&paymentRequestId=PAYR_REJ",
        )
        assertTrue(r is BushaPayResult.Cancelled)
        r as BushaPayResult.Cancelled
        assertEquals(BushaPayCancelledReason.REJECTED, r.reason)
        assertEquals("PAYR_REJ", r.paymentId)
    }

    @Test
    fun parseCallback_failedYieldsErrorWithExplicitCodeAndMessage() {
        val r = parseCallback(
            "co.x.busha-pay://callback?status=failed" +
                "&error_code=card_declined&error_message=Card%20declined",
        )
        assertTrue(r is BushaPayResult.Error)
        r as BushaPayResult.Error
        assertEquals("card_declined", r.code)
        assertEquals("Card declined", r.message)
    }

    @Test
    fun parseCallback_fallsBackToStatusAsErrorCode() {
        val r = parseCallback("co.x.busha-pay://callback?status=failed") as BushaPayResult.Error
        assertEquals("failed", r.code)
        assertEquals("Payment failed", r.message)
    }

    @Test
    fun parseCallback_usesUnknownWhenStatusMissing() {
        val r = parseCallback("co.x.busha-pay://callback") as BushaPayResult.Error
        assertEquals("unknown", r.code)
    }
}
